dvc/remote/base.py

```diff
@@ -275,13 +299,55 @@ def get_checksum(self, path_info):
             checksum = self.get_file_checksum(path_info)
 
         if checksum:
-            self.state.save(path_info, checksum)
+            self.state.save(path_info, checksum, checksum_type)
 
         return checksum
 
     def save_info(self, path_info):
         assert path_info.scheme == self.scheme
-        return {self.PARAM_CHECKSUM: self.get_checksum(path_info)}
+        return {self.checksum_type(): self.get_checksum(path_info)}
+
+    def changed_checksum(self, path_info, checksum_info, addition_check=None):
+        if not checksum_info:
+            logger.debug(
+                "checksum for '{}' is missing({}).".format(
+                    path_info, checksum_info
+                )
+            )
+            return True
+
+        types = set(checksum_info.keys())
+        supported_types = set(self.checksum_types())
+        intersection = supported_types.intersection(types)
+        if not supported_types.intersection(types):
+            logger.debug(
+                "checksum for '{}' is not supported({}).".format(
+                    path_info, checksum_info
+                )
+            )
+            return True
+
+        for typ in intersection:
+            expected = checksum_info.get(typ)
+            if addition_check:
+                if addition_check(expected, typ):
+                    logger.debug(
+                        "{} for '{}'('{}') has changed.".format(
+                            addition_check, path_info, checksum_info
+                        )
+                    )
+                    return True
```

Comment:
Is this `addition_check` only purpose to pass there `self.changed_cache`? If so won't it be more clear to simple use bool parameter?